### ISSUE TITLE
Auto-source config files in `$nu.default-config-dir/conf.d` and modules from `$nu.default-config-dir/modules.d`

### ISSUE THREAD

**Issue by NotTheDr01ds** [nushell/nushell#11918]
### Related problem

It's a great feature from Fish that I truly think could elevate Nushell as well.

At present, configuration changes pretty much always require modification of the default `config.nu`.  This is way too much like POSIX shells that I left behind several years ago for greener pastures.  It eventually leads to large, monolithic config files that are error-prone and difficult to troubleshoot.

It's also difficult for third-party code to be added to the configuration in an idempotent manner.  The end-user must always modify the config, rather than having a package installer do it automatically.

For instance, the current Starship prompt installation instructions for Nu are:

> Add the following to the end of your Nushell env file (find it by running `$nu.env-path` in Nushell):
>
> ```
> mkdir ~/.cache/starship
> starship init nu | save -f ~/.cache/starship/init.nu
> ```
>
> And add the following to the end of your Nushell configuration (find it by running $nu.config-path):
>
> ```
> use ~/.cache/starship/init.nu
> ```

It would be extremely difficult for Starship (or any third-party package) to make that change on the user's behalf, since it involves finding and replacing any previous invocations, regardless of slight differences in those commands.  For this reason, most software packages just tell the user to "make the change themselves".

### Describe the solution you'd like

Imagine if installing Starship (or most third-party packages) into Nushell was as easy as a **one-time**:

```Nushell
starship init nu | save -f $"($nu.default-config-dir)/modules.d/starship.nu"
```

This would require a (I believe) fairly simple feature where:

* Any files in `~/.config/nushell/conf.d` (adjusted for `config-path`) would automatically be sourced after `config.nu` is processed (or perhaps `login.nu`?)
* Any modules in `~/.config/nushell/modules.d` would be automatically imported

Ordering can be handled by filename sort-order -- It's not ideal, but it's worked for me for many years on Fish.  It also follows most other `conf.d` implementations (e.g. Apache, IIRC).

### Describe alternatives you've considered

Writing my own config-builder tool, which would create a `config.nu` that imported/sourced the files in those directories.

### Additional context and details

Note that Starship's instructions for Fish could be improved to use this method.  Far too many users seem to be stuck in the painful-POSIX-past, unfortunately, so they default to the "just modify the config file yourself".

If we get out in front of this, implement a similar feature, and promote/educate its use, then I think we can tout Nushell as being far simpler to configure than other shells (well, after perhaps some other changes that I'll require later 😉).

**Comment by fdncred** [comment]
We've talked about this recently in some other issue that I can't find. The synopsis, as I remember it, was to have something like NU_LIB_DIRS where script files/modules could be stored and if they are present get sourced on startup. I probably won't support folder named `conf.d` and `modules.d` because they're way to Linux-y. However, if there was something like NU_AUTO_SOURCE, and it was a list of folders, you could name them anything you want.

**Comment by ADIX7** [comment]
Is there an ETA on this? This is one of the most important thing missing for me to switch to nushell. Managing config files in 3rd party tools (part of dotfiles) would be much more easy.

**Comment by NotTheDr01ds** [comment]
@ADIX7 I'm not a Rust developer (yet, but I'd love to learn), but it seems to me to be a relatively easy implementation.  Might be a "Good First Issue" for someone to work on, but I'll let a Nu team member make that determination.

**Comment by NotTheDr01ds** [comment]
And I came back over to just to add that this isn't just for third-party modules.

It's far more easy to just `<nano/vi/edit/code> ~/.config/nushell/conf.d/my_function.nu` and edit away than it is to find the function/custom command declaration in a monolithic `config.nu`, even with the best of IDE/editors.

**Comment by NotTheDr01ds** [comment]
Throwing out a ... maybe ... include subdirectories of `conf.d`/`modules.d` automatically as well?  This would allow you to `git clone` a module/script directly into the directory ...

**Comment by fdncred** [comment]
> Is there an ETA on this? This is one of the most important thing missing for me to switch to nushell. Managing config files in 3rd party tools (part of dotfiles) would be much more easy.

It's just waiting on someone to implement it.

**Comment by fdncred** [comment]
> Good first issue for someone, or are there nuances I'm not anticipating?

Yes, there will be nuances like:
1. when should they be loaded, before x file, after y file
2. when should they be loaded, with x parameter or with y parameter
3. should all of these be merged into the root overlay or should it be a new one
4. probably other concerns here too

**Comment by NotTheDr01ds** [comment]
@fdncred  Fair enough

I'll try to experiment with various flavors of this (via hard-coded `source` and `use` statements in `config.nu` for now) to see if I can flesh out the recommended implementation.

For #1, at least, I recommend loading after all default configs are processed (`env.nu`, `config.nu`, and `login.nu` - Is that the complete set?).  This allows the `conf.d` files to act as overrides on any "default" or pre-existing settings.  Again, that's the way other `conf.d` implementations tend to go.

The main question on item 1, though, that I'm not sure about is how to choose the ordering of modules vs. `source`'d files.  Again, I'll experiment.

**Comment by fdncred** [comment]
> I recommend loading after all default configs are processed

i agree

number 2 is about `nu --parameters` e.g. should it be loaded with nu -c, what about nu -c -l, what about nu --stdin, what about....

>  that I'm not sure about is how to choose the ordering of modules vs. source'd files. 

One thing about ordering, iirc, some linux types do 1.filename 2.filename 3.filename, etc. i'm not sure we need to do that or not and i'm also not sure how that plays with source, source-env, use, overlay, overlay use, etc

**Comment by lpnh** [comment]
If I've understood correctly, this sounds like a great feature to be implemented!

Could this be related to the following?  #5682 #8214

**Comment by fdncred** [comment]
@NotTheDr01ds 

The first small part
![image](https://github.com/nushell/nushell/assets/343840/cf2a016a-94e5-480d-a9f3-466c14e15c7c)

**Comment by NotTheDr01ds** [comment]
@fdncred I still haven't thought of a "good" name for the autoload "config".  Right now, for my implementation, I have:

* `$nu.default_config_dir/env`: All `nu` files in it are auto-sourced via `env.nu`.
* `$nu.default_config_dir/config`:  Everything in it is auto-sourced via `config.nu`.
* `$nu.default_config_dir/modules`:  Everything in it is auto-`use`'d via `config.nu`.
* `$nu.default_config_dir/keybindings`:  Each file contains a keybinding `record` that gets appended (via `++=`) into `$env.config.keybindings`

Then, not necessarily related to the auto-loads, I have the following directories that are added to `$env.NU_LIB_DIRS`:

* `$nu.default_config_dir/lib/modules`
* `$nu.default_config_dir/lib/scripts`

It's the `$nu.default_config_dir/config` directory that I can't quite "like".  I mean, it's *all* *"config"*.  I could call it `auto-source`, but I don't quite like that either.  `scripts` seems a bit off, but perhaps not.  I tend to think of *"scripts"* as something that gets run via `nu <scriptfile>`.

**Comment by NotTheDr01ds** [comment]
Closed with #14669.
